cpp/src/arrow/filesystem/hdfs.cc

```diff
@@ -69,6 +69,14 @@ class HadoopFileSystem::Impl {
   HdfsOptions options() const { return options_; }
 
   Result<FileInfo> GetFileInfo(const std::string& path) {
+    // It has unfortunately been a frequent logic error to pass URIs down
+    // to GetFileInfo (e.g. ARROW-10264).  Unlike other filesystems, HDFS
+    // silently accepts URIs but returns different results than if given the
+    // equivalent in-filesystem paths.  Instead of raising cryptic errors
+    // later, notify the underlying problem immediately.
+    if (path.substr(0, 5) == "hdfs:") {
```

Comment:
or "viewfs" ? 
(I am not familiar with it, I only know that in the python/cython code there are some places that checks for this as well ..)

Reply:
I know, though I don't want to goldplate this either.